### PR TITLE
DM-45873: Initial query_all_datasets implementation

### PR DESCRIPTION
**Pull request by dhirving**
## Checklist

Extracted the query logic from the `query-datasets` CLI and used it to provide an initial implementation of query_all_datasets.  

This gets the interface and unit tests in place for a future PR that will run the whole query on the server side for RemoteButler.

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

### CODE REVIEW

File: python/lsst/daf/butler/script/queryDatasets.py [hunk 1]
```diff
@@ -258,83 +260,50 @@ def getDatasets(self) -> Iterator[list[DatasetRef]]:
...
+        else:
+            limit = self._limit
+
+        try:
+            pages = query_all_datasets(
+                self.butler,
+                collections=query_collections,
                 find_first=self._find_first,
+                name=self._dataset_type_glob,
                 with_dimension_records=True,
```
Comment (timj) [inline python/lsst/daf/butler/script/queryDatasets.py:283]: Following on from a Slack comment, it looks like there are other scripts using the `QueryDatasets` class that do need dimensions so if you want to be able to turn this off for `query-datasets` but turn it on for `transfer-datasets` then you will need to add a parameter higher up.

File: tests/test_cliCmdQueryDatasets.py [hunk 2]
```diff
@@ -357,6 +357,13 @@ def test_limit_order(self):
         ]
         self.assertAstropyTablesEqual(tables, expectedTables, filterColumns=True)
 
+        # Same as previous test, but with positive limit so no warning is
+        # issued.
+        tables = self._queryDatasets(
+            repo=testRepo.butler, limit=1, order_by=("visit"), collections="*", glob="*"
```
Comment (timj) [inline tests/test_cliCmdQueryDatasets.py]: Maybe test that no warning was issued (by issuing your own warning and then making sure you only got the one warning)?

**Review by timj — APPROVED** [review]
Thanks. I've tested it and it does seem to work as expected. It looks like the restriction of order-by only working for a single dataset type is not part of this PR yet.

File: python/lsst/daf/butler/script/queryDatasets.py (hunk 1 above)
Comment (dhirving) [reply python/lsst/daf/butler/script/queryDatasets.py:283]: Thanks -- I failed to notice that `QueryDatasets` was shared by multiple scripts.  I had removed order-by and the dimension records in the follow-up PR to this one but I'll have to put the dimension records back.

**Pull request merged**